### ISSUE TITLE
Hama - Signature Replay Attack Vulnerability in claimBySignature Function

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-06-tokensoft-judging#123]
Hama

high

# Signature Replay Attack Vulnerability in claimBySignature Function

## Summary
The claimBySignature function in the smart contract is vulnerable to a signature replay attack. The vulnerability arises due to the lack of nonce usage or other mechanisms to prevent the reuse of the same signature, allowing potential attackers to replay a valid signature on different transactions and claim tokens on behalf of the same beneficiary multiple times.

## Vulnerability Detail
The claimBySignature function allows a beneficiary to claim tokens using a merkle proof and beneficiary signature. However, the vulnerability lies in the signature verification process. The signature is recovered from the provided data and then checked to ensure it corresponds to the provided beneficiary address. This step is necessary to confirm that the beneficiary is authorized to claim tokens. However, the contract does not incorporate any mechanism to prevent the reuse of the same signature, leaving it susceptible to a signature replay attack.

## Impact
The signature replay vulnerability can have severe consequences for the smart contract and its users. Potential impacts include:

Unauthorized Token Transfers: Malicious actors can intercept a valid signature and replay it to repeatedly claim tokens on behalf of the same beneficiary, leading to unauthorized token transfers and a loss of funds.

Depletion of Token Reserves: With repeated signature replays, the smart contract's token reserves can be significantly depleted, affecting the contract's ability to fulfill legitimate claims from other beneficiaries.

Trust Erosion: The presence of this vulnerability may undermine the trust of users in the contract, affecting its reputation and adoption.

## Code Snippet
https://github.com/sherlock-audit/2023-06-tokensoft/blob/main/contracts/contracts/claim/abstract/CrosschainMerkleDistributor.sol#L119

      function claimBySignature(
    address _recipient,
    uint32 _recipientDomain,
    address _beneficiary,
    uint32 _beneficiaryDomain,
    uint256 _total,
    bytes calldata _signature,
    bytes32[] calldata _proof
    ) external {
    // Recover the signature by beneficiary
    bytes32 _signed = keccak256(
      abi.encodePacked(_recipient, _recipientDomain, _beneficiary, _beneficiaryDomain, _total)
    );
    address recovered = _recoverSignature(_signed, _signature);
    require(recovered == _beneficiary, '!recovered');

    // Validate the claim
    _verifyMembership(_getLeaf(_beneficiary, _total, _beneficiaryDomain), _proof);
    uint256 claimedAmount = _executeClaim(_beneficiary, _total);

    _settleClaim(_beneficiary, _recipient, _recipientDomain, claimedAmount);
   }

## Tool used
Manual Review

## Recommendation
Nonce Usage: Introduce a nonce or unique identifier that is included in the data signed by the beneficiary. Maintain a mapping of used nonces to ensure each nonce is used only once, thereby preventing the reuse of the same signature.

Duplicate of #10
